Completion list for type literals in type arguments

tiagovtristao opened the pull request:
This PR aims to provide some type literal completion for type arguments, as suggested in https://github.com/microsoft/TypeScript/issues/33302. It takes in mind intersection and union types, as well as nested type literals.

DanielRosenwasser commented on src/services/completions.ts:
```diff
@@ -983,6 +988,58 @@ namespace ts.Completions {
         return !!symbol.declarations?.some(d => d.kind === SyntaxKind.SourceFile);
     }
 
+    function getTypeLiteralInTypeArgumentCompletions(sourceFile: SourceFile, position: number, contextToken: Node | undefined, checker: TypeChecker, options: CompilerOptions, log: Log, preferences: UserPreferences): CompletionInfo | undefined{
```
Nit: maybe make this more obvious by calling it an `ObjectTypeLiteral`.

DanielRosenwasser commented on src/services/completions.ts:
```diff
@@ -983,6 +988,58 @@ namespace ts.Completions {
+
+        const typeLiteralNode = tryGetTypeLiteralNode(contextToken);
+
+        if (typeLiteralNode) {
```
You can reduce a level of nesting with an early return to `undefined` here.

DanielRosenwasser commented:
This looks like it's going in the right direction, but I haven't looked at this part of the codebase in a bit.

andrewbranch commented:
@typescript-bot pack this

DanielRosenwasser commented:
Annoying that the playground build didn't work.

@typescript-bot pack this

DanielRosenwasser reviewed (commented):
Looking at the code, it looks very clean and it seems like it does exactly what I'd expect - actually, it does more than that, specifically accounting for intersections which I hadn't thought of!

I guess one thing I do wonder about is how this interacts with symbols, and non-identifier keys (e.g. keys with spaces, keys with numbers at the beginning, fully numeric keys). Those are very niche scenarios so I won't exactly block on that, but completions shouldn't feel intrusive there.

We should also consider these to be new identifier definition locations (e.g. you can specify more than what's there without the editor getting in your way).

(see https://github.com/microsoft/TypeScript/issues/42595#issuecomment-771269621 for what I mean on that last point))

andrewbranch commented on src/services/completions.ts:
```diff
@@ -2857,6 +2909,47 @@ namespace ts.Completions {
+        return undefined;
+    }
+
+    function tryGetTypeArgumentSubType(node: Node, checker: TypeChecker): Type | undefined {
```
This name is a little confusing—it looks like it’s getting the constraint of a type argument, and then optionally drilling into a property of that constraint type, based on the syntax location of the cursor. I was confused about why we’d want to get a subtype of a type argument when its constraint is necessarily a supertype of it. Maybe `getConstraintOfTypeArgumentProperty` or something?

tiagovtristao replied:
Agreed the naming isn't correct, and I was looking for something like what you suggested. Nice!

andrewbranch reviewed (commented):
A couple nits, but looks great overall.

andrewbranch commented on src/services/completions.ts:
```diff
@@ -153,6 +153,11 @@ namespace ts.Completions {
             return stringCompletions;
         }
 
+        const objectTypeLiteralCompletions = getObjectTypeLiteralInTypeArgumentCompletions(sourceFile, position, contextToken, typeChecker, compilerOptions, log, preferences);
+        if (objectTypeLiteralCompletions) {
+            return objectTypeLiteralCompletions;
+        }
```
Instead of putting this here, did you look at adding to `getCompletionData` -> `tryGetGlobalSymbols`? It feels similar to `tryGetObjectLikeCompletionSymbols()` and `tryGetClassLikeCompletionSymbols()` so I expected it to be another case in that big chain of logical ORs.

tiagovtristao replied:
I did look into it initially, but wasn't sure whether it would be the best place to put it. However, I tried moving the code there, and it still works as expected, so I think it should be fine :-)

tiagovtristao commented:
> Looking at the code, it looks very clean and it seems like it does exactly what I'd expect - actually, it does more than that, specifically accounting for intersections which I hadn't thought of!
> 
> I guess one thing I do wonder about is how this interacts with symbols, and non-identifier keys (e.g. keys with spaces, keys with numbers at the beginning, fully numeric keys). Those are very niche scenarios so I won't exactly block on that, but completions shouldn't feel intrusive there.
> 
> We should also consider these to be new identifier definition locations (e.g. you can specify more than what's there without the editor getting in your way).
> 
> (see [#42595 (comment)](https://github.com/microsoft/TypeScript/issues/42595#issuecomment-771269621) for what I mean on that last point)

- I just added support for non-identifier keys. Good catch!
- I'm not sure I fully understood what you meant by "how this interacts with symbols" but I tried 2 things: Attempting to support JS symbols as keys but found out this isn't currently supported by TS as per https://github.com/Microsoft/TypeScript/issues/1863; And guaranteeing that keys that share the same escaped names of available symbols are supported, to which I added extra test checks.
- Marked these as new identifier definition locations as suggested

DanielRosenwasser commented:
@typescript-bot pack this

andrewbranch reviewed (approved):
Looks good to me, thanks! @DanielRosenwasser, any other edge cases you want to check?
